lib/github/hooks.nit

```diff
@@ -0,0 +1,172 @@
+#			print "new member {event.member.login}"
+#		else if event isa PullRequestEvent then
+#			print "new pr event on #{event.number}"
+#		else if event isa PullRequestReviewCommentEvent then
+#			print "new pr comment on #{event.pull.number}"
+#		else if event isa PushEvent then
+#			print "new push event from {event.ref}"
+#		else if event isa StatusEvent then
+#			print "new status {event.state} for {event.commit.sha}"
+#		else
+#			print "unknwon event..."
+#		end
+#	end
+# end
+#
+# var api = new GithubAPI(get_github_oauth)
+# var listener = new LogHookListener(api, "127.0.0.1", 8080)
+# # listener.listen # uncomment to start listening
+# ~~~
+module hooks
+
+import events
+import nitcorn
+
+# A nitcorn listener for Github hooks.
+class HookListener
+
+	# Api client used to perform Github API requests.
+	var api: GithubAPI
+
+	# Host to listen.
+	var host: String
+
+	# Port to listen.
+	var port: Int
+
+	# VirtualHost listened
+	private var vh: VirtualHost is noinit
+
+	init do
+		vh = new VirtualHost("{host}:{port}")
+		vh.routes.add new Route("/", new HookAction(self))
+	end
+
+	# Verbosity level (0: quiet, 1: debug).
+	# Default: 0
+	var verbosity = 0
+
+	# Print `message` if `lvl` <= `verbosity`
+	fun message(lvl: Int, message: String) do
+		if lvl <= verbosity then print message
+	end
+
+	# Start listening and responding to event.
+	fun listen do
+		message(1, "Hook listening on {host}:{port}")
+		var factory = new HttpFactory.and_libevent
+		factory.config.virtual_hosts.add vh
+		factory.run
+	end
+
+	# How to build events from received json objects.
+	fun event_factory(kind: String, json: JsonObject): GithubEvent do
+		if kind == "commit_comment" then
+			return new CommitCommentEvent.from_json(api, json)
+		else if kind == "create" then
+			return new CreateEvent.from_json(api, json)
+		else if kind == "delete" then
+			return new DeleteEvent.from_json(api, json)
+		else if kind == "deployment" then
+			return new DeploymentEvent.from_json(api, json)
+		else if kind == "deployment_status" then
+			return new DeploymentStatusEvent.from_json(api, json)
+		else if kind == "fork" then
+			return new ForkEvent.from_json(api, json)
+		else if kind == "issues" then
+			return new IssuesEvent.from_json(api, json)
+		else if kind == "issue_comment" then
+			return new IssueCommentEvent.from_json(api, json)
+		else if kind == "member" then
+			return new MemberEvent.from_json(api, json)
+		else if kind == "pull_request" then
+			return new PullRequestEvent.from_json(api, json)
+		else if kind == "pull_request_review_comment" then
+			return new PullRequestReviewCommentEvent.from_json(api, json)
+		else if kind == "push" then
+			return new PushEvent.from_json(api, json)
+		else if kind == "status" then
+			return new StatusEvent.from_json(api, json)
+		else
+			return new GithubEvent.from_json(api, json)
+		end
+	end
+
+	# What to do when we receive an event from a hook?
+	#
+	# This method MUST be defined in specialized listeners.
```

Comment:
La doc générée devrait déjà indiquer le `abstract`, non ? Forcer le programmeur à écrire dans sa doc une phrase qui veut dire exactement `abstract` n'est pas digne d'une documentation parfaite. Ou alors `abstract` n'est pas exactement synonyme de `This method MUST be defined in specialized subclasses` mais dans ce cas là on a un problème de spec et il faut définir mieux que ce que `abstract` veut dire au niveau du langage.

Le fait que le compilo soit taquin et laisse le programme se faire compiler et exécuter jusqu'à faire boom au dernier moment boom est un problème indépendant #808